docs/RFCS/sql_batching.md

```diff
@@ -0,0 +1,669 @@
+read and write-sets for our pending batch can simply hold table IDs. The `sql.Executor`
+will iterate over provided statements one at a time, as it does now in
+`execStmtsInCurrentTxn`. For each subsequent statement, the Executor will first check
+if it is one of the [statement types that we support](#batchable-statements-types) for
+statement batching. If not, the previous batch of statements will execute and the current
+statement will execute after they are done.
+
+If the current statement is supported through statement batching, we first create its `planNode`
+and then collect the [tables](#dependency-granularity) that it reads and writes to. This will
+be accomplished using a new `DependencyCollector` type that will implement the `planObserver`
+interface and traverse over the statement's `planNode`. The read and write-sets collected from
+the plan will then be compared against the current batch's read and write-sets using the rules for
+[independent statements](#independent-statements). If the statement is deemed to be independent
+of the current batch of statements, then it is added to the current batch and its read and
+write-sets are merged with the current batch's. If the statement is deemed dependent on the
+current batch of statements, then the current batch is [executed in parallel](#sqlkv-interaction).
+Once this completes, the batch and its read and write-sets are all cleared and the new statement is
+added as the only member of the current batch, with its read and write-sets replacing the
+batch's. Note that because no DDL statement types will be batchable, the planNode for this
+statement will not need to be reinitialized.
+
+The integration of this algorithm in the `sql.Executor` needs to be adjusted slightly if we
+decide to support batching in a [conversational API](#conversational-api-support), but the
+general algorithm remains the same.
+
+## Batchable Statements Types
+
+Initially, there will be **5** types of statements that we will look to parallelize
+when possible:
+- SELECT *
+- INSERT
+- UPDATE
+- DELETE
+- UPSERT **
+
+These statements make up the majority of use cases where statement batching would be
+useful, and provide fairly straightforward semantics about read and write-sets. Another
+important quality of these statements is that none of them mutate SQL schemas, meaning that
+the execution or lack of execution of one of these statements will never necessitate the
+re-initialization of the `planNode` for another. _Is this true even if we allow UPDATEs on
+system tables?_
+
+\* SELECT statements are included because they should be relatively easy to support and fit
+into the model we've laid out above. However, if this turns out not to be the case, it is ok
+to remove them initially as they are not an integral part of the targeted workflow.
+
+\*\* As mentioned [below](#coalesced-batches), UPSERT statements may not be easily supported
+using coalesced batches.
+
```

Comment:
I think I understand what you're saying, and it sounds a lot like @andreimatei's suggestion at the end of [this comment](https://github.com/cockroachdb/cockroach/pull/13160#issuecomment-277795897). My last change added this as an alternative (see `Span-Based Dependency Detection`). While I really like the idea in theory, I voiced a few concerns above:

> My main issue with an approach like this is that we'd be mixing up two fundamentally different concepts, those being SQL data structures and their underlying key-value representations. The problem with this is that this statement independence analysis would then become dependent on the details of our SQL->KV mapping. This means that future changes to this mapping, changes to the KV access patterns by the SQL layer, or even options like column families or interleaved tables could all subtly change the semantics here. In essence, we'd be leaking the details of the KV-layer into the behavior of this new feature. I really like the elegance of this approach, but I think the logical mismatch between the two concepts is just large enough for it to end up causing more complications than it solves.